File: go/cmd/network-rtt/udp.go
package main

import (
	"bytes"
	"fmt"
	"net"
	"time"
)

// udpReadTimeout is the per-recv deadline. Loopback UDP is effectively
// lossless; a timeout is treated as a hard error, not a retransmit.
const udpReadTimeout = time.Second

// measureUDP starts an in-process UDP echo server on loopback, dials one
// connected client socket, runs warmup then cfg.Iterations strict ping-pong
// round trips, and returns the per-round-trip elapsed nanoseconds.
func measureUDP(cfg Config) ([]int64, error) {
	srvAddr, err := net.ResolveUDPAddr("udp", "127.0.0.1:0")
	if err != nil {
		return nil, fmt.Errorf("udp: resolve: %w", err)
	}
	srvConn, err := net.ListenUDP("udp", srvAddr)
	if err != nil {
		return nil, fmt.Errorf("udp: listen: %w", err)
	}
	defer srvConn.Close()

	srvErr := make(chan error, 1)
	stop := make(chan struct{})
	go udpEchoServer(srvConn, cfg.PayloadBytes, stop, srvErr)

	// DialUDP connects the socket so plain Read/Write work.
	cliConn, err := net.DialUDP("udp", nil, srvConn.LocalAddr().(*net.UDPAddr))
	if err != nil {
		return nil, fmt.Errorf("udp: dial: %w", err)
	}
	defer cliConn.Close()

	send := make([]byte, cfg.PayloadBytes)
	for i := range send {
		send[i] = byte(i)
	}
	recv := make([]byte, cfg.PayloadBytes)

	roundTrip := func() error {
		if _, err := cliConn.Write(send); err != nil {
			return fmt.Errorf("udp: write: %w", err)
		}
		if err := cliConn.SetReadDeadline(time.Now().Add(udpReadTimeout)); err != nil {
			return fmt.Errorf("udp: set deadline: %w", err)
		}
		n, err := cliConn.Read(recv)
		if err != nil {
			return fmt.Errorf("udp: read: %w", err)
		}
		if n != len(send) || !bytes.Equal(recv[:n], send) {
			return fmt.Errorf("udp: echo mismatch")
		}
		return nil
	}

	// Warmup (discarded).
	for i := 0; i < cfg.Warmup; i++ {
		if err := roundTrip(); err != nil {
			return nil, err
		}
	}

	// Measured.
	samples := make([]int64, cfg.Iterations)
	for i := 0; i < cfg.Iterations; i++ {
		start := time.Now()
		if err := roundTrip(); err != nil {
			return nil, err
		}
		samples[i] = time.Since(start).Nanoseconds()
	}

	// Stop the server and surface any server-side error.
	close(stop)
	srvConn.Close()
	if err := <-srvErr; err != nil {
		return nil, err
	}
	return samples, nil
}

// udpEchoServer reads datagrams and echoes them back to the sender until
// stopped (signalled by closing stop, then closing the conn).
func udpEchoServer(conn *net.UDPConn, payloadBytes int, stop <-chan struct{}, errc chan<- error) {
	buf := make([]byte, payloadBytes)
	for {
		n, addr, err := conn.ReadFromUDP(buf)
		if err != nil {
			select {
			case <-stop:
				errc <- nil // expected shutdown
			default:
				errc <- fmt.Errorf("udp echo: read: %w", err)
			}
			return
		}
		if _, err := conn.WriteToUDP(buf[:n], addr); err != nil {
			select {
			case <-stop:
				errc <- nil
			default:
				errc <- fmt.Errorf("udp echo: write: %w", err)
			}
			return
		}
	}
}

File: rust/network-rtt/src/udp.rs
//! UDP synchronous ping-pong RTT measurement over loopback.
//!
//! An in-process echo server runs on its own thread; the client `connect()`s
//! its socket and issues one datagram at a time, timing each round trip. A read
//! timeout guards against (unexpected) loopback loss — a timeout is a hard
//! error, never a retransmit.

use std::io::{self};
use std::net::UdpSocket;
use std::sync::mpsc;
use std::thread;
use std::time::{Duration, Instant};

use crate::config::Config;

/// Run the UDP benchmark and return one elapsed-nanosecond sample per measured
/// round trip (`cfg.iterations` of them).
pub fn run(cfg: &Config) -> io::Result<Vec<u64>> {
    let server_sock = UdpSocket::bind("127.0.0.1:0")?;
    let server_addr = server_sock.local_addr()?;
    let payload_bytes = cfg.payload_bytes;

    // Signal used to stop the echo server once the client is done.
    let (done_tx, done_rx) = mpsc::channel::<()>();

    // Echo server: bounce every datagram back to its sender. A short read
    // timeout lets the loop check the shutdown signal without blocking forever.
    let server = thread::spawn(move || -> io::Result<()> {
        server_sock.set_read_timeout(Some(Duration::from_millis(200)))?;
        let mut buf = vec![0u8; payload_bytes];
        loop {
            match server_sock.recv_from(&mut buf) {
                Ok((n, src)) => {
                    server_sock.send_to(&buf[..n], src)?;
                }
                Err(e)
                    if e.kind() == io::ErrorKind::WouldBlock
                        || e.kind() == io::ErrorKind::TimedOut =>
                {
                    // Timed out waiting for a datagram; check for shutdown.
                    if done_rx.try_recv().is_ok() {
                        return Ok(());
                    }
                }
                Err(e) => return Err(e),
            }
        }
    });

    let result = client_loop(server_addr, cfg);

    // Tell the server to stop and wait for it.
    let _ = done_tx.send(());
    server
        .join()
        .map_err(|_| io::Error::new(io::ErrorKind::Other, "udp echo server thread panicked"))??;

    result
}

/// Connect, warm up, then measure `cfg.iterations` round trips.
fn client_loop(server_addr: std::net::SocketAddr, cfg: &Config) -> io::Result<Vec<u64>> {
    let sock = UdpSocket::bind("127.0.0.1:0")?;
    sock.connect(server_addr)?;
    // Timeout = hard error (see module docs); 1s per the design.
    sock.set_read_timeout(Some(Duration::from_secs(1)))?;

    let send = vec![0xCDu8; cfg.payload_bytes];
    let mut recv = vec![0u8; cfg.payload_bytes];

    for _ in 0..cfg.warmup {
        round_trip(&sock, &send, &mut recv)?;
    }

    let mut samples = vec![0u64; cfg.iterations];
    for slot in samples.iter_mut() {
        let start = Instant::now();
        round_trip(&sock, &send, &mut recv)?;
        *slot = start.elapsed().as_nanos() as u64;
    }

    Ok(samples)
}

/// One ping-pong datagram: send the payload, recv the echo, assert equality.
#[inline]
fn round_trip(sock: &UdpSocket, send: &[u8], recv: &mut [u8]) -> io::Result<()> {
    sock.send(send)?;
    let n = sock.recv(recv).map_err(|e| {
        if e.kind() == io::ErrorKind::WouldBlock || e.kind() == io::ErrorKind::TimedOut {
            io::Error::new(io::ErrorKind::TimedOut, "udp recv timed out (loopback loss)")
        } else {
            e
        }
    })?;
    if &recv[..n] != send {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            "udp echo mismatch: received bytes differ from sent",
        ));
    }
    Ok(())
}
